### ISSUE TITLE
Recursive query execution: LLM-driven iterative search via code generation

### ISSUE THREAD

**Issue by anibjoshi** [stratalab/strata-core#1272]
# Recursive Query Execution

## Summary

Integrate the [Recursive Language Model (RLM)](https://github.com/alexzhang13/rlm) paradigm into Strata's search and inference pipeline. Instead of single-shot retrieval, let the LLM **iteratively explore the database through code generation** — issuing searches, traversing graphs, inspecting results, and recursively decomposing sub-problems — until it converges on an answer.

This builds on top of the Graph-Augmented Hybrid Search pipeline (#1270) as the ultimate expression of inference-embedded-in-the-database: the query engine itself reasons recursively over Strata's primitives.

### What are RLMs?

Recursive Language Models ([Zhang, Kraska, Khattab — MIT, arXiv:2512.24601](https://arxiv.org/abs/2512.24601)) are an inference-time orchestration paradigm where:

1. Context is stored externally (not stuffed into the prompt)
2. The LLM writes code to explore, slice, grep, and analyze that context
3. The LLM can recursively call itself (`rlm_query()`) on sub-problems
4. Each recursive call gets its own execution scope with a subset of context

Results: RLM(GPT-5-mini) outperformed vanilla GPT-5 by 114% on long-context tasks. An 8B model fine-tuned for recursion (RLM-Qwen3-8B) approached GPT-5 quality. The approach handles 10M+ token contexts — two orders of magnitude beyond model context windows.

### Why Strata is the ideal host for this

RLM's reference implementation uses a Python REPL where the model greps strings and chunks text. Strata replaces all of that with **indexed, searchable, graph-connected primitives**:

| RLM's REPL Approach | Strata-Native Equivalent |
|---|---|
| `context[0:2000]` (string slicing) | `db.search("keywords", k=10)` (BM25 + vector) |
| `re.findall(pattern, context)` (regex grep) | `db.search("pattern", mode="keyword")` |
| `chunks = [context[i:i+N] for i in ...]` (manual chunking) | Already chunked as KV/JSON entries |
| `llm_query("Summarize: " + chunk)` (sub-LM call) | `db.generate(model, prompt)` (in-process) |
| Store results in Python dict | `db.kv_set("intermediate_result", value)` (persisted) |
| No structure awareness | Full graph + ontology traversal |

The model doesn't need to grep a string — it can query a database. And that database has BM25, vectors, a knowledge graph, and an ontology.

#### Tier 1: Iterative Search (minimal, no code execution)

The model issues a search, sees results, decides to refine, and issues follow-up searches in a loop. No REPL, no code generation — just prompt chaining.

```
Iteration 1:
  LLM: Search for "metformin side effects"
  → db.search("metformin side effects", graph_aware=True, k=20)
  → 20 results returned to LLM context

Iteration 2:
  LLM: Results mention GI issues frequently. Narrow to elderly patients.
  → db.search("metformin GI side effects elderly", anchor_nodes=["age_group_60+"], k=10)
  → 10 refined results

Iteration 3:
  LLM: Found the answer in results 2 and 7. FINAL(answer).
```

**Implementation**: A loop in the search handler that calls `Generate` → parses a structured action (search/traverse/answer) → executes → feeds results back. Uses existing `Generate` and `Search` commands. ~300 lines.

**Termination**: Max iterations (default 5), max token budget, or model signals `FINAL`.

#### Tier 2: Database-Native RLM (code execution over Strata API)

The model writes code that calls Strata's API directly. Instead of a Python REPL over raw strings, it's a scripted interaction with the database.

```python
# Model-generated code executed in sandbox
results = db.search("metformin side effects", k=50)
patient_nodes = db.graph_neighbors("medical", "medication-metformin", "incoming", "prescribed")
elderly = [n for n in patient_nodes if db.graph_get_node("medical", n)["properties"]["age"] > 60]
notes = []
for p in elderly[:20]:
    refs = db.graph_neighbors("medical", p, "outgoing", "authored")
    notes.extend(refs)
# Recursive sub-call on the filtered notes
summaries = rlm_batch([f"Summarize side effects in: {db.kv_get(n)}" for n in notes[:10]])
FINAL_VAR(summaries)
```

**Implementation**: Embed a lightweight script interpreter (Rhai, Lua, or PyO3) that exposes Strata's API. The model generates code blocks, the interpreter executes them against the live database, results feed back into the model's context.

**Strata API surface exposed to the sandbox**:
- `db.search(query, ...)` — hybrid search
- `db.kv_get(key)` / `db.json_get(id)` — read primitives
- `db.graph_neighbors(graph, node, direction, edge_type)` — graph traversal
- `db.graph_nodes_by_type(graph, type)` — type-based lookup
- `db.graph_ontology_summary(graph)` — schema introspection
- `db.generate(model, prompt)` — sub-LM call
- `rlm_query(prompt, context)` — recursive self-call with new scope
- `FINAL(answer)` / `FINAL_VAR(var)` — termination

**NOT exposed** (safety): writes, deletes, config changes, branch operations. The sandbox is read-only + inference-only.

**Estimated scope**: ~800 lines for the execution loop + sandbox + API bindings.

#### Tier 3: Branch-Scoped Recursive Exploration

Each recursive call gets its own branch for isolated exploration state:

```
Root (branch: main):
  "Analyze all clinical notes for metformin side effects by age group"
  
  → Branch "rlm-explore-1": analyze notes for age 40-60
  → Branch "rlm-explore-2": analyze notes for age 60-80
  → Branch "rlm-explore-3": analyze notes for age 80+
  
  Each branch can write intermediate results (kv_set) without
  affecting main. Root reads from all branches, merges findings.
  Branches discarded after query completes.
```

This uses Strata's copy-on-write branching — each recursive call is an isolated exploration with zero overhead until it writes.

**Estimated scope**: Builds on Tier 2 + ~200 lines for branch lifecycle management.

## Resource Control

RLMs have unbounded cost/latency by default. Strata needs strict resource controls:

```rust
pub struct RecursiveQueryConfig {
    /// Maximum iterations at the root level (default: 10)
    pub max_iterations: usize,
    /// Maximum recursion depth (default: 1)  
    pub max_depth: usize,
    /// Maximum total LLM tokens across all calls (default: 50,000)
    pub max_total_tokens: usize,
    /// Maximum wall-clock time (default: 30s)
    pub max_timeout: Duration,
    /// Maximum concurrent sub-calls (default: 4)
    pub max_concurrent: usize,
}
```

## Relationship to RFC #1270

This issue extends the Graph-Augmented Hybrid Search RFC (#1270):

- **RFC Phase 5 (LLM Retrieval Planning)** generates a plan and executes it in a single pass. Recursive query execution lets the model **iterate on the plan** based on intermediate results.
- **RFC Phases 1-4** (graph boost, context reranking, ontology expansion, fusion) become **tools the recursive model can invoke** — not a fixed pipeline, but a toolkit.
- **RFC branching addendum** directly enables Tier 3's branch-scoped exploration.

The progression: Phase 5 is "the LLM plans the search." Recursive execution is "the LLM **runs** the search, inspects results, and adapts."

## Implementation Plan

| Phase | What | Complexity | Depends On |
|---|---|---|---|
| Tier 1: Iterative search | Prompt-chain loop with Generate + Search | Low (~300 lines) | Nothing — can start now |
| Tier 2: Sandbox execution | Embedded interpreter + Strata API bindings | Medium (~800 lines) | Tier 1 proven useful |
| Tier 3: Branch-scoped recursion | Copy-on-write branches per recursive call | Low add-on (~200 lines) | Tier 2 |
| Fine-tuned model | SFT a small model (8B) on Strata-native RLM trajectories | High (research) | Tier 2 generating training data |

## Open Questions

1. **Interpreter choice**: Rhai (pure Rust, safe, limited stdlib) vs Lua (battle-tested, fast) vs PyO3 (full Python, matches RLM reference impl but heavy dependency). Rhai is the most "embedded-friendly" option.

2. **Training data**: RLM-Qwen3-8B was fine-tuned on Python REPL trajectories. A Strata-native model would need training data using Strata's API instead. Tier 2 could generate this data organically.

3. **GGUF compatibility**: RLM-Qwen3-8B is not yet available in GGUF format. When it is, it could run via strata-inference's llama.cpp backend. Until then, Tier 1 works with any model.

4. **Determinism**: Recursive queries are inherently non-deterministic (the model chooses different exploration paths). Should we support `seed` parameters for reproducibility?

5. **Observability**: Each recursive call should emit structured traces (iteration number, depth, action taken, tokens used, latency) for debugging and cost monitoring.

## References

- [Recursive Language Models (blog)](https://alexzhang13.github.io/blog/2025/rlm/)
- [arXiv:2512.24601](https://arxiv.org/abs/2512.24601) — Zhang, Kraska, Khattab (MIT)
- [github.com/alexzhang13/rlm](https://github.com/alexzhang13/rlm) — Reference implementation
- [RLM-Qwen3-8B](https://huggingface.co/alexzhang13/RLM-Qwen3-8B-Base) — Fine-tuned 8B model
- RFC #1270 — Graph-Augmented Hybrid Search
